### DISCUSSION TITLE
Using hover and focus variants when not disabled

### DISCUSSION THREAD

**Discussion by amiranagram** [tailwindlabs/tailwindcss#3574] [Help] [answered]
**I want to apply classes on hover and focus only if the button is not disabled.**

What approach should I take?

Should I customize variants so it would only apply styles when `:not([disabled])`, and how would I do that?
Or is there any way I could have a not variant? I could use it like `not:disabled:hover:bg-blue-700` or `enabled:hover:bg-blue-700`.

**Comment by simonswiss** [comment, marked as answer]
Hey!

**In short:** you can do the opposite of what you want: Add disabled styles if the button is disabled. These will "cancel" any hover/focus styles you have applied on the same element.

---

**Detailed answer**

You can enable the [`disabled` variant](https://tailwindcss.com/docs/hover-focus-and-other-states#disabled) for what you need (like `backgroundColor, textColor`) in your config file's `variants`.

If a `disabled:` style is set for a button, it will never allow hover/focus styles to show. Here's an example in Tailwind Play:

https://play.tailwindcss.com/LOaudRbGXs

Notice how I enabled the `disabled` variant for both the background and text colors, and then applied `disabled` styles to both buttons (one of them having the `disabled` attribute).

![image](https://user-images.githubusercontent.com/485747/107893343-bcdb8e80-6f7e-11eb-9344-d33a884bb130.png)

Hope it helps 👍

**Reply by amiranagram** [reply]
This is great! Appreciate it.

**Reply by hyunbinseo** [reply]
I know this is an old discussion, but Tailwind v4 now supports [`not-*` variant](https://tailwindcss.com/blog/tailwindcss-v4#not-variant)

Instead of setting `disabled:` styles to override, just use `not-disabled:`

```diff
- hover:bg-indigo-400 disabled:bg-gray-200
+ not-disabled:bg-indigo-400
```

This is how I use it:

```
bg-slate-500 not-disabled:hover:bg-slate-600 disabled:opacity-30
```

**Comment by rafahuaman** [comment]
@simonswiss Thanks for all your work as a maintainer!

I found some strange behavior that differs from the detailed answer above. I was wondering if it was a bug or expected behavior.
> If a disabled: style is set for a button, it will never allow hover/focus styles to show

The `hover` styles are still applied even when the following classes are the only disabled styles applied `disabled:opacity-50 disabled:cursor-not-allowed`. If I add another class like `bg-red-100`, the disabled component works as expected (hover does not trigger hover styles).

https://play.tailwindcss.com/ACFyBZ5QUV

**Reply by prince272** [reply]
Any solution yet?

**Reply by adamwathan** [reply]
If you only want the hover styles to show when the button is enabled, add `enabled:`:

`hover:enabled:bg-indigo-400`

https://play.tailwindcss.com/Af6OyarFsl

**Reply by prince272** [reply]
Thanks. I never thought about that.

**Reply by bauhaas** [reply]
Nice little tips

**Reply by selemondev** [reply]
This solution doesn't seem to work for dynamically generated classnames. How can I go about this?

**Reply by ctoma** [reply]
`hover:enabled:bg-indigo-400` works but beware if you have say `active:bg-indigo-300` then the hover will always override it, I guess due to specificity. 
To fix this you need to then declare your active class with enabled also, i.e. `active:enabled:bg-indigo-300`

**Comment by sesa736123** [comment]
@selemondev From my knowledge dynamically generated classes doesn't work at all in Tailwind

**Comment by simoami** [comment]
I used the `enabled:hover:*` before but It has no effect on anchor elements. You end up losing the hover. Fyi, an anchor is sometimes used in place of a button to render links with href. If you're using radix-ui through shadcn/ui, this would be the equivalent of `<Button asChild><Link href="...">A link</Link></Button>` which renders a link tag `<a href="">...</a>`. 
One solution is to stick with a plain `hover:*` directive combined with `disabled:hover:<revert hover styles>`, where if the disabled attribute is not applicable, the hover remains in effect. Alternatively, you can also just add `disabled:pointer-events-none` next to a simple hover rule if you don't need any pointer or keyboard detection.

**Reply by Nases** [reply]
Exactly our problem, we use same class for both button & a, but enabled breaks hover effects for anchor tags.
